diff -rup mutter-original/src/compositor/mutter-plugin.c mutter-mod/src/compositor/mutter-plugin.c
--- mutter-original/src/compositor/mutter-plugin.c	2009-10-02 15:50:14.000000000 -0400
+++ mutter-mod/src/compositor/mutter-plugin.c	2010-01-21 11:01:59.780585475 -0500
@@ -33,6 +33,7 @@
 
 #include "compositor-private.h"
 #include "mutter-window-private.h"
+#include "mutter-window-group.h"
 
 G_DEFINE_ABSTRACT_TYPE (MutterPlugin, mutter_plugin, G_TYPE_OBJECT);
 
@@ -352,6 +353,48 @@ mutter_plugin_get_window_group (MutterPl
   return mutter_get_window_group_for_screen (priv->screen);
 }
 
+ClutterActor *
+mutter_plugin_get_below_window_group (MutterPlugin *plugin)
+{
+  ClutterActor *actor;
+  MutterWindowGroup *group;
+  MutterPluginPrivate *priv = MUTTER_PLUGIN (plugin)->priv;
+
+
+  actor = mutter_get_window_group_for_screen (priv->screen);
+  group = MUTTER_WINDOW_GROUP (actor);
+  
+  return mutter_window_group_get_below (group);
+}
+
+ClutterActor *
+mutter_plugin_get_normal_window_group (MutterPlugin *plugin)
+{
+  ClutterActor *actor;
+  MutterWindowGroup *group;
+  MutterPluginPrivate *priv = MUTTER_PLUGIN (plugin)->priv;
+
+
+  actor = mutter_get_window_group_for_screen (priv->screen);
+  group = MUTTER_WINDOW_GROUP (actor);
+  
+  return mutter_window_group_get_normal (group);
+}
+
+ClutterActor *
+mutter_plugin_get_above_window_group (MutterPlugin *plugin)
+{
+  ClutterActor *actor;
+  MutterWindowGroup *group;
+  MutterPluginPrivate *priv = MUTTER_PLUGIN (plugin)->priv;
+
+
+  actor = mutter_get_window_group_for_screen (priv->screen);
+  group = MUTTER_WINDOW_GROUP (actor);
+  
+  return mutter_window_group_get_above (group);
+}
+
 /**
  * _mutter_plugin_effect_started:
  * @plugin: the plugin
diff -rup mutter-original/src/compositor/mutter-window.c mutter-mod/src/compositor/mutter-window.c
--- mutter-original/src/compositor/mutter-window.c	2009-08-24 14:28:39.000000000 -0400
+++ mutter-mod/src/compositor/mutter-window.c	2010-01-19 13:04:47.469591342 -0500
@@ -19,6 +19,7 @@
 #include "compositor-private.h"
 #include "mutter-shaped-texture.h"
 #include "mutter-window-private.h"
+#include "mutter-window-group.h"
 #include "shadow.h"
 #include "tidy/tidy-texture-frame.h"
 
@@ -1201,8 +1202,7 @@ mutter_window_new (MetaWindow *window)
   /* Hang our compositor window state off the MetaWindow for fast retrieval */
   meta_window_set_compositor_private (window, G_OBJECT (self));
 
-  clutter_container_add_actor (CLUTTER_CONTAINER (info->window_group),
-			       CLUTTER_ACTOR (self));
+  mutter_window_group_pack_window (MUTTER_WINDOW_GROUP (info->window_group), self);
   clutter_actor_hide (CLUTTER_ACTOR (self));
 
   /* Initial position in the stack is arbitrary; stacking will be synced
diff -rup mutter-original/src/compositor/mutter-window-group.c mutter-mod/src/compositor/mutter-window-group.c
--- mutter-original/src/compositor/mutter-window-group.c	2009-08-19 10:46:43.000000000 -0400
+++ mutter-mod/src/compositor/mutter-window-group.c	2010-01-19 21:34:45.310788603 -0500
@@ -15,11 +15,72 @@ struct _MutterWindowGroup
 {
   ClutterGroup parent;
 
-  MetaScreen *screen;
+  MetaScreen   *screen;
+  ClutterActor *above;
+  ClutterActor *normal;
+  ClutterActor *below;
 };
 
 G_DEFINE_TYPE (MutterWindowGroup, mutter_window_group, CLUTTER_TYPE_GROUP);
 
+ClutterActor *
+mutter_window_group_get_above (MutterWindowGroup *group)
+{
+  g_return_val_if_fail (MUTTER_IS_WINDOW_GROUP (group), NULL);
+  
+  return group->above;
+}
+
+ClutterActor *
+mutter_window_group_get_normal (MutterWindowGroup *group)
+{
+  g_return_val_if_fail (MUTTER_IS_WINDOW_GROUP (group), NULL);
+  
+  return group->normal;
+}
+
+ClutterActor *
+mutter_window_group_get_below (MutterWindowGroup *group)
+{
+  g_return_val_if_fail (MUTTER_IS_WINDOW_GROUP (group), NULL);
+  
+  return group->below;
+}
+
+void
+mutter_window_group_pack_window (MutterWindowGroup *group, 
+                                 MutterWindow *window)
+{
+  MetaCompWindowType type;
+  ClutterContainer  *container;
+  
+  g_return_if_fail (MUTTER_IS_WINDOW_GROUP (group));
+  g_return_if_fail (MUTTER_IS_WINDOW (window));
+  
+  type = mutter_window_get_window_type (window);
+  
+  if (type == META_COMP_WINDOW_DROPDOWN_MENU ||
+      type == META_COMP_WINDOW_POPUP_MENU ||
+      type == META_COMP_WINDOW_TOOLTIP ||
+      type == META_COMP_WINDOW_NOTIFICATION ||
+      type == META_COMP_WINDOW_COMBO ||
+      type == META_COMP_WINDOW_DND ||
+      type == META_COMP_WINDOW_OVERRIDE_OTHER)
+    {
+      container = CLUTTER_CONTAINER (group->above);
+    }
+  /*else if (type == META_COMP_WINDOW_DESKTOP)
+    {
+      container = CLUTTER_CONTAINER (group->below);
+    }*/
+  else
+    {
+      container = CLUTTER_CONTAINER (group->normal);
+    }
+  
+  clutter_container_add_actor (container, CLUTTER_ACTOR (window));
+}
+
 /* We want to find out if the window is "close enough" to
  * 1:1 transform. We do that by converting the transformed coordinates
  * to 24.8 fixed-point before checking if they look right.
@@ -95,6 +156,22 @@ actor_is_untransformed (ClutterActor *ac
   return TRUE;
 }
 
+static GList *
+mutter_window_group_get_windows (MutterWindowGroup *window_group)
+{
+  GList *children, *other;
+  
+  children = clutter_container_get_children (CLUTTER_CONTAINER (window_group->below));
+  
+  other = clutter_container_get_children (CLUTTER_CONTAINER (window_group->normal));
+  children = g_list_concat (children, other);
+  
+  other = clutter_container_get_children (CLUTTER_CONTAINER (window_group->above));
+  children = g_list_concat (children, other);
+  
+  return children;
+}
+
 static void
 mutter_window_group_paint (ClutterActor *actor)
 {
@@ -107,7 +184,7 @@ mutter_window_group_paint (ClutterActor 
    * and subtract the opaque area of each window out of the visible
    * region that we pass to the windows below.
    */
-  children = clutter_container_get_children (CLUTTER_CONTAINER (actor));
+  children = mutter_window_group_get_windows (window_group);
   children = g_list_reverse (children);
 
   /* Start off with the full screen area (for a multihead setup, we
@@ -179,6 +256,17 @@ mutter_window_group_class_init (MutterWi
 static void
 mutter_window_group_init (MutterWindowGroup *window_group)
 {
+  window_group->below  = clutter_group_new ();
+  clutter_container_add_actor (CLUTTER_CONTAINER (window_group), window_group->below);
+  
+  window_group->normal = clutter_group_new ();
+  clutter_container_add_actor (CLUTTER_CONTAINER (window_group), window_group->normal);
+  
+  window_group->above  = clutter_group_new ();
+  clutter_container_add_actor (CLUTTER_CONTAINER (window_group), window_group->above);
+  
+  clutter_container_raise_child (CLUTTER_CONTAINER (window_group), window_group->normal, NULL);
+  clutter_container_raise_child (CLUTTER_CONTAINER (window_group), window_group->above, NULL);
 }
 
 ClutterActor *
diff -rup mutter-original/src/compositor/mutter-window-group.h mutter-mod/src/compositor/mutter-window-group.h
--- mutter-original/src/compositor/mutter-window-group.h	2009-08-05 17:53:31.000000000 -0400
+++ mutter-mod/src/compositor/mutter-window-group.h	2010-01-19 11:34:26.599587232 -0500
@@ -45,8 +45,16 @@ typedef struct _MutterWindowGroup       
 typedef struct _MutterWindowGroupClass   MutterWindowGroupClass;
 typedef struct _MutterWindowGroupPrivate MutterWindowGroupPrivate;
 
-GType mutter_window_group_get_type (void);
+GType         mutter_window_group_get_type    (void);
 
-ClutterActor *mutter_window_group_new (MetaScreen *screen);
+void          mutter_window_group_pack_window (MutterWindowGroup *group, MutterWindow *window);
+
+ClutterActor *mutter_window_group_get_above   (MutterWindowGroup *group);
+
+ClutterActor *mutter_window_group_get_normal  (MutterWindowGroup *group);
+
+ClutterActor *mutter_window_group_get_below   (MutterWindowGroup *group);
+
+ClutterActor *mutter_window_group_new         (MetaScreen *screen);
 
 #endif /* MUTTER_WINDOW_GROUP_H */
diff -rup mutter-original/src/include/mutter-plugin.h mutter-mod/src/include/mutter-plugin.h
--- mutter-original/src/include/mutter-plugin.h	2009-08-28 12:27:29.000000000 -0400
+++ mutter-mod/src/include/mutter-plugin.h	2010-01-21 10:59:49.481835295 -0500
@@ -236,6 +236,15 @@ ClutterActor *
 mutter_plugin_get_window_group (MutterPlugin *plugin);
 
 ClutterActor *
+mutter_plugin_get_above_window_group (MutterPlugin *plugin);
+
+ClutterActor *
+mutter_plugin_get_normal_window_group (MutterPlugin *plugin);
+
+ClutterActor *
+mutter_plugin_get_below_window_group (MutterPlugin *plugin);
+
+ClutterActor *
 mutter_plugin_get_stage (MutterPlugin *plugin);
 
 void
