=== modified file 'src/GeisAdapter.cpp'
--- a/src/GeisAdapter.cpp	2011-03-30 17:29:51 +0000
+++ b/src/GeisAdapter.cpp	2011-04-01 09:15:34 +0000
@@ -20,6 +20,10 @@
 #include <glib.h>
 #include "GeisAdapter.h"
 
+#define UNITY_GESTURE_PRIMITIVE_TOUCH 32
+#define UNITY_GESTURE_TYPE_TOUCH3     "Touch,touch=3"
+#define UNITY_GESTURE_TYPE_TOUCH4     "Touch,touch=4"
+
 GeisAdapter * GeisAdapter::_default = 0;
 
 /* static */
@@ -126,7 +130,7 @@
     self->tap.emit (data);
     g_free (data);
   }
-  else if (gesture_type == GEIS_GESTURE_PRIMITIVE_TOUCH)
+  else if (gesture_type == UNITY_GESTURE_PRIMITIVE_TOUCH)
   {
     GeisTouchData *data = self->ProcessTouchGesture (count, attrs);
     data->id = gesture_id;
@@ -168,7 +172,7 @@
     self->tap.emit (data);
     g_free (data);
   }
-  else if (gesture_type == GEIS_GESTURE_PRIMITIVE_TOUCH)
+  else if (gesture_type == UNITY_GESTURE_PRIMITIVE_TOUCH)
   {
     GeisTouchData *data = self->ProcessTouchGesture (count, attrs);
     data->id = gesture_id;
@@ -210,7 +214,7 @@
     self->tap.emit (data);
     g_free (data);
   }
-  else if (gesture_type == GEIS_GESTURE_PRIMITIVE_TOUCH)
+  else if (gesture_type == UNITY_GESTURE_PRIMITIVE_TOUCH)
   {
     GeisTouchData *data = self->ProcessTouchGesture (count, attrs);
     data->id = gesture_id;
@@ -428,8 +432,8 @@
 }
 
 static const char* s_gestures[] = {
-  GEIS_GESTURE_TYPE_DRAG3, GEIS_GESTURE_TYPE_TAP3, GEIS_GESTURE_TYPE_ROTATE3, GEIS_GESTURE_TYPE_PINCH3, GEIS_GESTURE_TYPE_TOUCH3,
-  GEIS_GESTURE_TYPE_DRAG4, GEIS_GESTURE_TYPE_TAP4, GEIS_GESTURE_TYPE_ROTATE4, GEIS_GESTURE_TYPE_PINCH4, GEIS_GESTURE_TYPE_TOUCH4,
+  GEIS_GESTURE_TYPE_DRAG3, GEIS_GESTURE_TYPE_TAP3, GEIS_GESTURE_TYPE_ROTATE3, GEIS_GESTURE_TYPE_PINCH3, UNITY_GESTURE_TYPE_TOUCH3,
+  GEIS_GESTURE_TYPE_DRAG4, GEIS_GESTURE_TYPE_TAP4, GEIS_GESTURE_TYPE_ROTATE4, GEIS_GESTURE_TYPE_PINCH4, UNITY_GESTURE_TYPE_TOUCH4,
   GEIS_GESTURE_TYPE_SYSTEM,
   NULL
 };

