diff -Nur -x '*.orig' -x '*~' unity-packaging//targets/mutter/main.c unity-packaging.new//targets/mutter/main.c
--- unity-packaging//targets/mutter/main.c	2010-05-27 16:41:15.180713000 +0200
+++ unity-packaging.new//targets/mutter/main.c	2010-06-07 12:42:02.071691919 +0200
@@ -92,10 +92,6 @@
 static gboolean unity_mutter_xevent_filter (MutterPlugin *self,
                                       XEvent      *event);
 
-static void unity_mutter_topmost_changed (MutterPlugin *plugin,
-                                          MutterWindow *old_window,
-                                          MutterWindow *new_window);
-
 static void on_restore_input_region (UnityPlugin *plugin, gboolean fullscreen);
 
 static const MutterPluginInfo * unity_mutter_plugin_info (MutterPlugin *self);
@@ -117,7 +113,6 @@
   mut_class->kill_effect      = unity_mutter_kill_effect;
   mut_class->xevent_filter    = unity_mutter_xevent_filter;
   mut_class->plugin_info      = unity_mutter_plugin_info;
-  mut_class->topmost_changed  = unity_mutter_topmost_changed;
 }
 
 static void
@@ -269,16 +264,6 @@
   unity_plugin_kill_effect (UNITY_MUTTER (self)->plugin, window, events);
 }
 
-static void
-unity_mutter_topmost_changed (MutterPlugin *self,
-                              MutterWindow *old_window,
-                              MutterWindow *new_window)
-{
-  unity_plugin_topmost_changed (UNITY_MUTTER (self)->plugin,
-                                old_window,
-                                new_window);
-}
-
 static gboolean
 unity_mutter_xevent_filter (MutterPlugin *self,
                             XEvent      *event)
diff -Nur -x '*.orig' -x '*~' unity-packaging//targets/mutter/plugin.vala unity-packaging.new//targets/mutter/plugin.vala
--- unity-packaging//targets/mutter/plugin.vala	2010-06-07 12:21:01.646437000 +0200
+++ unity-packaging.new//targets/mutter/plugin.vala	2010-06-07 12:42:02.071691919 +0200
@@ -834,30 +834,6 @@
       this.window_kill_effect (this, window, events);
     }
 
-    public void topmost_size_changed (Clutter.Actor           actor,
-                                      Clutter.ActorBox        box,
-                                      Clutter.AllocationFlags flags)
-    {
-      if (actor is Mutter.Window)
-        check_fullscreen_obstruction ();
-    }
-
-    public void topmost_changed (Mutter.Window old_window,
-                                 Mutter.Window new_window)
-    {
-      if (active_window is Mutter.Window)
-        active_window.allocation_changed.disconnect (topmost_size_changed);
-
-      active_window = new_window;
-
-      if (active_window is Mutter.Window)
-        {
-          active_window.allocation_changed.connect (topmost_size_changed);
-
-          check_fullscreen_obstruction ();
-        }
-    }
-
     public int get_panel_height ()
     {
       return this.PANEL_HEIGHT;
