Index: ubuntu/plugins/unityshell/src/unityshell.cpp
===================================================================
--- ubuntu.orig/plugins/unityshell/src/unityshell.cpp	2012-02-17 11:09:56.892606000 +0100
+++ ubuntu/plugins/unityshell/src/unityshell.cpp	2012-02-23 15:12:32.023137576 +0100
@@ -2616,8 +2616,8 @@
   // Workspaces
   std::string const workspaces = _("Workspaces");
   hints_.push_back(new shortcut::Hint(workspaces, "", "", _("Spread workspaces."), shortcut::COMPIZ_KEY_OPTION, "expo", "expo_key"));
-  hints_.push_back(new shortcut::Hint(workspaces, "", "", _("Switch workspaces."), shortcut::HARDCODED_OPTION, _("Control + Alt + Cursor Keys")));
-  hints_.push_back(new shortcut::Hint(workspaces, "", "", _("Move focused window to different workspace."), shortcut::HARDCODED_OPTION, _("Control + Alt + Shift + Cursor Keys")));
+  hints_.push_back(new shortcut::Hint(workspaces, "", "", _("Switch workspaces."), shortcut::HARDCODED_OPTION, _("Super + Shift + Cursor Keys")));
+  hints_.push_back(new shortcut::Hint(workspaces, "", "", _("Move focused window to different workspace."), shortcut::HARDCODED_OPTION, _("Super + Alt + Cursor Keys")));
 
   // Windows
   std::string const windows = _("Windows");
