Index: ubuntu/plugins/unityshell/src/unityshell.cpp
===================================================================
--- ubuntu.orig/plugins/unityshell/src/unityshell.cpp	2012-03-13 09:39:19.944369000 +0100
+++ ubuntu/plugins/unityshell/src/unityshell.cpp	2012-03-13 09:44:21.501069418 +0100
@@ -111,7 +111,6 @@
   , _in_paint(false)
   , relayoutSourceId(0)
   , _redraw_handle(0)
-  , alt_tap_timeout_id_(0)
   , newFocusedWindow(nullptr)
   , doShellRepaint(false)
   , allowWindowPaint(false)
@@ -372,9 +371,6 @@
   if (relayoutSourceId != 0)
     g_source_remove(relayoutSourceId);
 
-  if (alt_tap_timeout_id_)
-    g_source_remove(alt_tap_timeout_id_);
-
   ::unity::ui::IconRenderer::DestroyTextures();
   QuicklistManager::Destroy();
 
@@ -1608,12 +1604,6 @@
   if (!grab_index_)
     return false;
 
-  if (alt_tap_timeout_id_)
-  {
-    g_source_remove(alt_tap_timeout_id_);
-    alt_tap_timeout_id_ = 0;
-  }
-
   screen->addAction(&optionGetAltTabRight());
   screen->addAction(&optionGetAltTabDetailStart());
   screen->addAction(&optionGetAltTabDetailStop());
@@ -1808,19 +1798,6 @@
     action->setState(action->state() | CompAction::StateTermKey);
   last_hud_show_time_ = g_get_monotonic_time();
 
-  /* Workaround to fix #943194 */
-  alt_tap_timeout_id_ = g_timeout_add(local::ALT_TAP_DURATION, [] (gpointer data) -> gboolean {
-    auto self = static_cast<UnityScreen*>(data);
-
-    if (!self->switcher_controller_->Visible())
-    {
-      XUngrabKeyboard(self->screen->dpy(), CurrentTime);
-    }
-
-    self->alt_tap_timeout_id_ = 0;
-    return FALSE;
-  }, this);
-
   // pass key through
   return false;
 }
@@ -1840,12 +1817,6 @@
   if (!(state & CompAction::StateTermTapped))
     return false;
 
-  if (alt_tap_timeout_id_)
-  {
-    g_source_remove(alt_tap_timeout_id_);
-    alt_tap_timeout_id_ = 0;
-  }
-
   gint64 current_time = g_get_monotonic_time();
   if (current_time - last_hud_show_time_ > (local::ALT_TAP_DURATION * 1000))
   {
Index: ubuntu/plugins/unityshell/src/unityshell.h
===================================================================
--- ubuntu.orig/plugins/unityshell/src/unityshell.h	2012-03-12 10:18:17.351155000 +0100
+++ ubuntu/plugins/unityshell/src/unityshell.h	2012-03-13 09:44:36.989146215 +0100
@@ -290,7 +290,6 @@
   bool                                  _in_paint;
   guint32                               relayoutSourceId;
   guint32                               _redraw_handle;
-  guint32                               alt_tap_timeout_id_;
   typedef std::shared_ptr<CompAction> CompActionPtr;
   typedef std::vector<CompActionPtr> ShortcutActions;
   ShortcutActions _shortcut_actions;
