diff -Nur -x '*.orig' -x '*~' ubuntu//unity-private/launcher/application-controller.vala ubuntu.new//unity-private/launcher/application-controller.vala
--- ubuntu//unity-private/launcher/application-controller.vala	2010-08-02 13:59:27.525546000 +0200
+++ ubuntu.new//unity-private/launcher/application-controller.vala	2010-08-02 14:05:12.047896000 +0200
@@ -459,7 +459,7 @@
 
       try
         {
-          name = desktop_keyfile.get_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_NAME);
+          name = desktop_keyfile.get_locale_string (KeyFileDesktop.GROUP, KeyFileDesktop.KEY_NAME);
         }
       catch (Error e)
         {
@@ -468,7 +468,7 @@
 
       try
         {
-          name = desktop_keyfile.get_string (KeyFileDesktop.GROUP, "X-GNOME-FullName");
+          name = desktop_keyfile.get_locale_string (KeyFileDesktop.GROUP, "X-GNOME-FullName");
         }
       catch (Error e)
         {
