=== modified file 'configure.ac'
--- configure.ac	2010-06-04 17:03:39 +0000
+++ configure.ac	2010-06-07 11:07:36 +0000
@@ -53,7 +53,7 @@
 dnl Check for perceptualdiff
 AC_CHECK_PROG(HAVE_IMAGEDIFF, perceptualdiff, true, false)
 if test "x${HAVE_IMAGEDIFF}" = "xfalse"; then
-  AC_MSG_ERROR([*** `perceptualdiff` not found, required for UI tests])
+  AC_MSG_WARN([*** `perceptualdiff` not found, required for UI tests])
 fi
 
 dnl Setup a few consts

